cdap-encryption-ext-tink/src/main/java/io/cdap/cdap/encryption/tink/CloudKmsClient.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.encryption.tink;
+
+import com.google.api.gax.rpc.ApiException;
+import com.google.api.gax.rpc.StatusCode;
+import com.google.cloud.kms.v1.CryptoKey;
+import com.google.cloud.kms.v1.CryptoKey.CryptoKeyPurpose;
+import com.google.cloud.kms.v1.CryptoKeyName;
+import com.google.cloud.kms.v1.CryptoKeyVersion.CryptoKeyVersionAlgorithm;
+import com.google.cloud.kms.v1.CryptoKeyVersionTemplate;
+import com.google.cloud.kms.v1.KeyManagementServiceClient;
+import com.google.cloud.kms.v1.KeyRing;
+import com.google.cloud.kms.v1.KeyRingName;
+import com.google.cloud.kms.v1.LocationName;
+import com.google.common.io.CharStreams;
+import com.google.protobuf.Duration;
+import com.google.protobuf.Timestamp;
+import io.cdap.cdap.security.spi.encryption.CipherInitializationException;
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.time.temporal.ChronoUnit;
+import java.util.HashSet;
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Client wrapping Cloud KMS.
+ */
+public class CloudKmsClient implements Closeable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CloudKmsClient.class);
+
+  private final KeyManagementServiceClient client;
+  private final String projectId;
+  private final String location;
+  private final String keyRingId;
+  // In-memory cache to hold created crypto keys, this is to avoid checking if a given crypto key exists.
+  private final Set<String> knownCryptoKeys;
+
+  /**
+   * Constructs Cloud KMS client.
+   *
+   * @throws IOException if cloud kms client can not be created
+   */
+  CloudKmsClient(String projectId, String location, String keyRingId) throws IOException {
+    this.projectId = projectId;
+    this.location = location;
+    this.keyRingId = keyRingId;
+    this.client = KeyManagementServiceClient.create();
+    this.knownCryptoKeys = new HashSet<>();
+  }
+
+  /**
+   * Get project id from the metadata server. Makes a request to the metadata server that lives on
+   * the VM, as described at https://cloud.google.com/compute/docs/storing-retrieving-metadata.
+   */
+  private String getSystemProjectId(String metadataServerApi) throws IOException {
```

Comment:
any reason we need to do this manually instead of ServiceOptions.getDefaultProjectId()?

That way would also work when CDAP is running in non-GCP environments (like sandbox)

Reply:
This is remnants of another class I copied this class from in the Secure Store extension. Good call though, I went ahead and added the logic to read the project ID using `ServiceOptions.getDefaultProjectId()` if it's not set in CConf.